lib/Feature/FeatureModelParser.cpp

```diff
@@ -233,4 +234,318 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+          return false;
+        }
+      }
+    }
+  }
+  return true;
+}
+
+bool FeatureModelSxfmParser::parseFeatureTree(xmlChar *FeatureTree) {
+  // Split the lines of the feature tree by new lines
+  {
+    std::stringstream Ss(reinterpret_cast<const char *>(FeatureTree));
+    std::string To;
+    string Name;
+    bool Opt;
+    int LastIndentationLevel = -1;
+    int RootIndentation = -1;
+    int OrGroupCounter = 0;
+    std::map<int, string> IndentationToParentMapping;
+
+    // This map is used for the or group mapping
+    // Each entry represents an or group as a tuple where the first value is
+    // the name of the parent, the second is the relationship kind, and the
+    // third a vector consisting of the name of the children
+    std::map<int, std::tuple<string, Relationship::RelationshipKind, std::vector<string>>> OrGroupMapping;
+
+    if (FeatureTree == nullptr) {
+      std::cerr << "Failed to read in feature tree. Is it empty?" << std::endl;
+      return false;
+    }
+
+    while (std::getline(Ss, To)) {
+      Opt = false;
+
+      if (To.empty() || std::all_of(To.begin(),To.end(),isspace)) {
+        continue;
+      }
+
+      // For every line, count the indentation
+      // not more than 1 additional indentations are allowed to the original one
+      // However, we may have arbitrarily less indentations
+      if ((To.find(':', 0 )) == std::string::npos) {
+        std::cerr << "Colon is missing in line" << To << std::endl;
+        return false;
+      }
+
+      llvm::StringRef ToStringRef = llvm::StringRef(To);
+      llvm::StringRef IndentationString = ToStringRef.substr(0, ToStringRef.find(':'));
+      int CurrentIndentationLevel = IndentationString.count(Indentation);
+      int Diff = CurrentIndentationLevel - LastIndentationLevel;
+
+      // Remember the root indentation for later checks
+      if (LastIndentationLevel == -1) {
+        RootIndentation = CurrentIndentationLevel;
+      }
+
+      if ((LastIndentationLevel != -1) && Diff > 1) {
+        std::cerr << "Indentation error in feature tree in line " << To << std::endl;
+        return false;
+      }
+
+      // Move pointer to first character after indentation
+      // The first character has to be a colon followed by the type of
+      // the feature (m for mandatory, o for optional, a for alternative)
+      std::string::size_type Pos = CurrentIndentationLevel * Indentation.length() + 2;
+      std::optional<std::tuple<int, int>> Cardinalities;
+
+      switch (To.at(Pos - 1)) {
+      case 'o':
+        // Code for optional
+        Opt = true;
+        break;
+      case 'g':
+        // Code for an or group with different cardinalities
+        Opt = false;
+        // Extract the cardinality
+        Cardinalities = extractCardinality(To);
+        if (!Cardinalities.has_value()) {
+          return false;
+        }
+        break;
+      case ' ':
+        // Code for alternative child
+        Pos--;
+        break;
+      }
```

Comment:
What happens here if we get a different letter? Does the Dtd prevent this, so can we be sure that we won't see anything else except for these three, or should we add some `default` label with error handling here?

Reply:
You are right. I definitely missed that. Thanks!